Server statistics tasks MR

Mohammad-Nayef opened the pull request:
To use it just modify the path of appsettings.json in GlobalConfigurations folder

Mohammad-Nayef committed:
Added project and solution files

Mohammad-Nayef committed:
Added appsettings.json

Mohammad-Nayef committed:
Added server statistics class

Mohammad-Nayef committed:
Added PerformanceCounter package

Mohammad-Nayef committed:
Included the cpu usage

Mohammad-Nayef committed:
Included the available memory

Mohammad-Nayef committed:
Included memory usage

Mohammad-Nayef committed:
Created a property for the current server statistics

Mohammad-Nayef committed:
Added RabbitMQ to the solution

Mohammad-Nayef committed:
Added mongoDB

Mohammad-Nayef committed:
Added SignalR Project

Mohammad-Nayef committed:
Added an alert consumer

Mohammad-Nayef committed:
Added an alert sender

Mohammad-Nayef committed:
Added a project for handling the global app configurations

Mohammad-Nayef committed:
Created an extension class for detecting anomalies

Mohammad-Nayef committed:
Created an extension class for mapping server statistics models

Mohammad-Nayef committed:
Created an abstraction for RabbitMQ connection

Mohammad-Nayef committed:
Created a class to send messages through a message queue by topic

Mohammad-Nayef committed:
Created a consumer for the messages by topic

Mohammad-Nayef committed:
Added the main method of the project

Mohammad-Nayef committed:
Created interface for recieving messages

Mohammad-Nayef committed:
Created interface for sending messages

Mohammad-Nayef committed:
Moved SignalR sender and consumer to SignalREndpoint

Mohammad-Nayef committed:
Created interface for consuming alerts

Mohammad-Nayef committed:
Created interface for sending alerts

Mohammad-Nayef committed:
Created a repository for server statistics

Mohammad-Nayef committed:
Created models for server statistics

Mohammad-Nayef committed:
Created a service for storing statistics in the database

Mohammad-Nayef committed:
Created a service for publishing statistics to the message queue

Mohammad-Nayef committed:
Added main method for consuming alerts

Mohammad-Nayef committed:
Removing moved files

Mohammad-Nayef committed:
Modified project files

Mohammad-Nayef committed:
Reconnecting of connection failed

Mohammad-Nayef committed:
Bug fix

Mohammad-Nayef committed:
Handling an exception

Mohammad-Nayef committed:
Implementing IMessageSender

Mohammad-Nayef committed:
Added a using statement

Mohammad-Nayef committed:
Ignoring front end stuff

Mohammad-Nayef committed:
Removed front end stuff

RandaZraik commented on MessageQueue/ServerStatisticsReceiver.cs:
```diff
@@ -0,0 +1,104 @@
+            _channel.BasicConsume(_queueName, false, consumer);
+        }
+
+        private ServerStatisticsDTO? GetServerStatistics(BasicDeliverEventArgs args)
```
you can move this to an extension method

Mohammad-Nayef replied:
done

RandaZraik commented on MessageQueue/ServerStatisticsReceiver.cs:
```diff
@@ -0,0 +1,104 @@
+            return _serverStatistics.IncludeServerIdentifier(serverIdentifier);
+        }
+
+        private async Task DetectAndReportAnomaliesAsync()
```
I think both of these methods should be moved to a separate service, this receiver doesn't have to know that much of details.

RandaZraik replied:
I see that you maintain `_previousStatistics` in this class, but still feels like this logic needs to be moved out

Mohammad-Nayef replied:
done

RandaZraik commented on ServerStatisticsPublisher/Program.cs:
```diff
@@ -0,0 +1,45 @@
+    {
+        messageSender = new TopicMessageSender(
+            "Message sending",
+            "Statistics exchange",
```
these constants such queue names would be better to be moved to a different class in this project, same thing for the other constants in the other projects

Mohammad-Nayef replied:
done

RandaZraik commented on GlobalConfigurations/AppConfigurations.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace GlobalConfigurations
+{
+    public class AppConfigurations
```
I see this used across multiple projects, since config file cannot be changed on runtime, you can just read it once  from the json file and use it in the whole program lifetime. So you can declare a static object of this class in program.cs of each project, and then pass it to the class constructors to be used.

`AnomaliesDetection` would need to be changed to a class with a constructor that accepts this object for example.

Mohammad-Nayef replied:
done

RandaZraik commented on MessageQueue/ServerStatisticsReceiver.cs:
```diff
@@ -0,0 +1,104 @@
+            IServerStatisticsService statisticsService,
+            IAlertSender alertSender) : base(connectionName)
+        {
+            _channel.ExchangeDeclare(exchangeName, ExchangeType.Topic);
```
seems like these three lines here in the sender class are common and can be just moved to `RabbitMqConnector` constructor.

Mohammad-Nayef replied:
done

Mohammad-Nayef committed:
Moved getting server statistics to an extension method

Mohammad-Nayef committed:
Code cleanup

Mohammad-Nayef committed:
Moved 3 duplicated lines of code to the base class

Mohammad-Nayef committed:
Now previous statistics variable is updated in all cases

Mohammad-Nayef committed:
Used only one instance of the configurations for each project

Mohammad-Nayef committed:
Ignored more front end stuff

Mohammad-Nayef committed:
Moved server anomalies detection to a new service

Mohammad-Nayef committed:
Moved getting the server identifier to an extension method

Mohammad-Nayef committed:
Overrode ToString()

Mohammad-Nayef committed:
Bug fix

Used decimal percentage instead of 0 - 100 range
Dealing with it was considered as decimal percentage

Mohammad-Nayef committed:
Removed the config private instance

Mohammad-Nayef committed:
Used server anomalies service

Mohammad-Nayef committed:
Send and print the same statistics instead of new ones

Mohammad-Nayef committed:
Used configurations instead of hard coded values for RabbitMQ

Mohammad-Nayef committed:
Code cleanup

Mohammad-Nayef committed:
Used configurations instead of hard coded values for MongoDb

Mohammad-Nayef committed:
Used configurations instead of hard coded values for SignalR
